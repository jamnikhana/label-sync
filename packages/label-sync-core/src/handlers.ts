import * as Octokit from '@octokit/rest'
import {
  Config,
  RepositoryConfig,
  GithubLabel,
  getGithubLabelsFromRepositoryConfig,
  getRepositoriesFromConfiguration,
  getRepositoryFromName,
  getRepostioryLabels,
  getLabelsDiff,
  addLabelsToRepository,
  updateLabelsInRepository,
  removeLabelsFromRepository,
} from './labels'

/**
 * Handlers
 */

export interface SyncOptions {
  dryRun: boolean
  githubToken: string
}

export interface SyncReport {
  config: Config
  options: SyncOptions
  successes: RepositorySyncSuccessReport[]
  errors: RepositorySyncErrorReport[]
}

export interface RepositorySyncSuccessReport {
  name: string
  config: RepositoryConfig
  additions: GithubLabel[]
  updates: GithubLabel[]
  removals: GithubLabel[]
}

export interface RepositorySyncErrorReport {
  name: string
  config: RepositoryConfig
  message: string
}

export type RepositorySyncReport =
  | {
      status: 'success'
      report: RepositorySyncSuccessReport
    }
  | {
      status: 'error'
      report: RepositorySyncErrorReport
    }

export async function handleSync(
  config: Config,
  options: SyncOptions,
): Promise<SyncReport> {
  /**
   * Authentication
   */

  const client = new Octokit()

  client.authenticate({
    type: 'app',
    token: options.githubToken,
  })

  /**
   * Repositories Sync
   */

  const repositories = getRepositoriesFromConfiguration(config)
  const repositoryHandlers = repositories.map(repository =>
    handleRepository(repository.name, repository.config),
  )

  const repositoryReports = await Promise.all(repositoryHandlers)

  const report = generateSyncReport(repositoryReports)

  return {
    config: config,
    options: options,
    successes: report.successes,
    errors: report.errors,
  }

  /**
   * Helper functions
   */
  async function handleRepository(
    name: string,
    config: RepositoryConfig,
  ): Promise<RepositorySyncReport> {
    const repository = getRepositoryFromName(name)

    if (!repository) {
      return {
        status: 'error',
        report: {
          name: name,
          config: config,
          message: `Cannot decode the provided repository name ${name}`,
        },
      }
    }

    /**
     * Labels
     */

    // Github
    const currentLabels = await getRepostioryLabels(client, repository)

    // Local
    const newLabels = getGithubLabelsFromRepositoryConfig(config)

    /**
     * Diff
     */

    const diff = getLabelsDiff(currentLabels, newLabels)

    /**
     * Sync
     */

    if (options.dryRun) {
      return {
        status: 'success',
        report: {
          name,
          config,
          additions: diff.add,
          updates: diff.update,
          removals: diff.remove,
        },
      }
    } else {
      /**
       * Sync
       */
      try {
        const additions = await addLabelsToRepository(
          client,
          diff.add,
          repository,
        )
        const updates = await updateLabelsInRepository(
          client,
          diff.update,
          repository,
        )
        const removals = config.strict
          ? await removeLabelsFromRepository(client, diff.remove, repository)
          : diff.remove

        return {
          status: 'success',
          report: {
            name,
            config,
            additions,
            updates,
            removals,
          },
        }
      } catch (err) {
        return {
          status: 'error',
          report: {
            name: name,
            config: config,
            message: err.message,
          },
        }
      }
    }
  }

  function generateSyncReport(
    repositoryReports: RepositorySyncReport[],
  ): {
    successes: RepositorySyncSuccessReport[]
    errors: RepositorySyncErrorReport[]
  } {
    const report = repositoryReports.reduce<{
      successes: RepositorySyncSuccessReport[]
      errors: RepositorySyncErrorReport[]
    }>(
      (acc, report) => {
        switch (report.status) {
          case 'success':
            return { ...acc, successes: [...acc.successes, report.report] }
          case 'error':
            return { ...acc, errors: [...acc.errors, report.report] }
        }
      },
      { successes: [], errors: [] },
    )

    return report
  }
}
