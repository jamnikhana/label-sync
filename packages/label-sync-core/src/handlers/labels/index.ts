import * as labels from './labels'
import * as reporter from './reporter'
import { handleLabelSync } from './sync'

export { handleLabelSync }
