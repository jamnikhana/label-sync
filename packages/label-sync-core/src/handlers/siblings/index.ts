import * as siblings from './siblings'
import { SiblingSyncReport } from './reporter'
import { handleSiblingSync } from './sync'

export { handleSiblingSync, SiblingSyncReport }
