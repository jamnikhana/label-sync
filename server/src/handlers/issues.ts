import { Octokit } from 'probot'

/**
 * Opens an issue with a prescribed title and body.
 *
 * @param octokit
 * @param owner
 * @param reports
 */
export async function handleIssue(
  octokit: Octokit,
  owner: string,
  repo: string,
  title: string,
  body: string,
): Promise<Octokit.IssuesCreateResponse> {
  return octokit.issues
    .create({
      repo: repo,
      owner: owner,
      title: title,
      body: body,
    })
    .then(({ data }) => data)
}
