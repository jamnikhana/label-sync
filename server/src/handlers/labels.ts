import { Octokit } from 'probot'

import { LSCRepository, LSCLabel } from '../configuration'
import * as maybe from '../data/maybe'
import { GithubLabel, getRepositoryLabels, isLabel } from '../github'
import { Dict } from '../utils'

export type LabelSyncReport =
  | {
      status: 'Success'
      owner: string
      repo: string
      additions: GithubLabel[]
      updates: GithubLabel[]
      removals: GithubLabel[]
      config: {
        strict: boolean
        labels: Dict<LSCLabel>
      }
    }
  | {
      status: 'Failure'
      repo: string
      owner: string
      message: string
      config: {
        strict: boolean
        labels: Dict<LSCLabel>
      }
    }

/**
 *
 * Performs a labels sync with configuration.
 *
 * @param client
 * @param repository
 * @param config
 * @param options
 */
export async function handleLabelSync(
  octokit: Octokit,
  owner: string,
  repo: string,
  { strict, labels }: LSCRepository,
  persist: boolean = false,
): Promise<LabelSyncReport> {
  /**
   * Label Sync handler firstly loads current labels from Github,
   * and new labels from local configuration.
   *
   * After that it generates a diff and creates, updates or deletes
   * label definitions in a particular repository.
   */
  const remoteLabels = await getRepositoryLabels(octokit, { repo, owner })
  const labelsDiff = maybe.andThen(remoteLabels, calculateDiff(labels))

  if (labelsDiff === null) {
    return {
      status: 'Failure',
      owner,
      repo,
      message: `Couldn't make a diff of labels.`,
      config: { labels, strict },
    }
  }

  const { added, changed, removed } = labelsDiff

  const [additions, updates, removals] = await Promise.all([
    addLabelsToRepository(octokit, { repo, owner }, added, persist),
    updateLabelsInRepository(octokit, { repo, owner }, changed, persist),
    removeLabelsFromRepository(
      octokit,
      { repo, owner },
      removed,
      strict && persist,
    ),
  ])

  return {
    status: 'Success',
    repo,
    owner,
    additions,
    updates,
    removals,
    config: {
      strict,
      labels,
    },
  }
}

/**
 *
 * Calculates the diff of labels.
 *
 * @param currentLabels
 * @param newLabels
 */
export function calculateDiff(
  config: LSCRepository['labels'],
): (
  labels: GithubLabel[],
) => {
  added: GithubLabel[]
  changed: GithubLabel[]
  removed: GithubLabel[]
  unchanged: GithubLabel[]
} {
  return labels => {
    const labelsNames = labels.map(l => l.name)

    const added = Object.keys(config)
      .filter(label => !labelsNames.includes(label))
      .map(hydrateLabel)
    const changed = Object.keys(config)
      .filter(label => {
        const persisted = labelsNames.includes(label)
        const changed = !labels.some(isLabel(hydrateLabel(label)))
        return persisted && changed
      })
      .map(hydrateLabel)
    const unchanged = Object.keys(config)
      .filter(label => labels.some(isLabel(hydrateLabel(label))))
      .map(hydrateLabel)
    const removed = labels.filter(label => !config.hasOwnProperty(label.name))

    return {
      added,
      changed,
      removed,
      unchanged,
    }
  }

  /* Helper functions */

  function hydrateLabel(name: string): GithubLabel {
    return { ...config[name], name, default: false }
  }
}

/**
 *
 * Create new labels in a repository.
 *
 * @param github
 * @param labels
 * @param repository
 */
export async function addLabelsToRepository(
  github: Octokit,
  { repo, owner }: { repo: string; owner: string },
  labels: GithubLabel[],
  persist: boolean,
): Promise<GithubLabel[]> {
  /* Return config on non persist */
  if (!persist) {
    return labels
  }

  /* Perform sync on persist. */
  const actions = labels.map(label => addLabelToRepository(label))
  return Promise.all(actions)

  /**
   * Helper functions
   */
  async function addLabelToRepository(
    label: GithubLabel,
  ): Promise<GithubLabel> {
    return github.issues
      .createLabel({
        owner: owner,
        repo: repo,
        name: label.name,
        description: label.description,
        color: label.color,
      })
      .then(res => res.data)
  }
}

/**
 *
 * Updates labels in repository.
 *
 * @param github
 * @param labels
 * @param repository
 */
export async function updateLabelsInRepository(
  github: Octokit,
  { repo, owner }: { repo: string; owner: string },
  labels: GithubLabel[],
  persist: boolean,
): Promise<GithubLabel[]> {
  /* Return config on non persist */
  if (!persist) {
    return labels
  }

  /* Update values on persist. */
  const actions = labels.map(label => updateLabelInRepository(label))
  return Promise.all(actions)

  /**
   * Helper functions
   */
  async function updateLabelInRepository(
    label: GithubLabel,
  ): Promise<GithubLabel> {
    return github.issues
      .updateLabel({
        current_name: label.name,
        owner: owner,
        repo: repo,
        name: label.name,
        description: label.description,
        color: label.color,
      })
      .then(res => res.data)
  }
}

/**
 *
 * Removes labels from repository.
 *
 * @param github
 * @param labels
 * @param repository
 */
export async function removeLabelsFromRepository(
  github: Octokit,
  { repo, owner }: { repo: string; owner: string },
  labels: GithubLabel[],
  persist: boolean,
): Promise<GithubLabel[]> {
  const actions = labels.map(label => removeLabelFromRepository(label))
  if (persist) await Promise.all(actions)

  return labels

  /**
   * Helper functions
   */
  async function removeLabelFromRepository(
    label: GithubLabel,
  ): Promise<Octokit.IssuesDeleteLabelParams> {
    return github.issues
      .deleteLabel({
        owner: owner,
        repo: repo,
        name: label.name,
      })
      .then(res => res.data)
  }
}
