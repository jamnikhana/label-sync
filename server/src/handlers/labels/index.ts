export {
  getLabelsDiff,
  addLabelsToRepository,
  updateLabelsInRepository,
  removeLabelsFromRepository,
} from './labels'
export { createTerminalReport, LabelSyncReport } from './reporter'
export { handleLabelSync, LabelSyncOptions } from './sync'
