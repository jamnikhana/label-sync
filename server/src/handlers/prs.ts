import { Octokit } from 'probot'

/**
 * Creates a comment on a dedicated pull request.
 * @param octokit
 * @param owner
 * @param repo
 * @param number
 * @param message
 */
export async function handlePRComment(
  octokit: Octokit,
  owner: string,
  repo: string,
  number: number,
  message: string,
): Promise<Octokit.IssuesCreateCommentResponse> {
  return octokit.issues
    .createComment({
      owner: owner,
      repo: repo,
      body: message,
      number: number,
    })
    .then(({ data }) => data)
}
