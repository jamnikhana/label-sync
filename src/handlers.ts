import * as Octokit from '@octokit/rest'
import {
  RepositoryConfig,
  GithubLabel,
  getRepostioryLabels,
  getGithubLabelsFromRepositoryConfig,
  getRepositoryFromName,
  getLabelsDiff,
  addLabelsToRepository,
  updateLabelsInRepository,
  removeLabelsFromRepository,
} from './labels'

/**
 * Handlers
 */

export interface RepositorySyncReport {
  name: string
  configuration: RepositoryConfig
  additions: GithubLabel[]
  updates: GithubLabel[]
  removals: GithubLabel[]
}

/**
 *
 * Handles report installation.
 *
 * @param client
 * @param name
 * @param config
 */
export async function handleRepository(
  client: Octokit,
  name: string,
  configuration: RepositoryConfig,
): Promise<RepositorySyncReport> {
  const repository = getRepositoryFromName(name)

  if (!repository) {
    throw new Error(`Cannot decode the provided repository name ${name}`)
  }

  /**
   * Labels
   */

  // Github
  const currentLabels = await getRepostioryLabels(client, repository)

  // Local
  const newLabels = getGithubLabelsFromRepositoryConfig(configuration)

  /**
   * Diff
   */

  const diff = getLabelsDiff(currentLabels, newLabels)

  /**
   * Sync
   */
  const additions = await addLabelsToRepository(client, diff.add, repository)
  const updates = await updateLabelsInRepository(
    client,
    diff.update,
    repository,
  )
  const removals = configuration.strict
    ? await removeLabelsFromRepository(client, diff.remove, repository)
    : []

  return {
    name,
    configuration,
    additions,
    updates,
    removals,
  }
}
