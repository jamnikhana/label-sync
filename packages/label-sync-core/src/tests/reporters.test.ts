import { generateSyncReport } from '../'
import { SyncReport } from '../handlers'

describe('Reporters', () => {
  test('correctly build report for organization sync', async () => {
    const report = generateSyncReport({
      config: {},
      options: {
        githubToken: '',
        dryRun: true,
      },
      successes: [
        {
          name: 'github-labels',
          config: {
            labels: {
              test: {
                description: 'description',
                color: 'color',
              },
              bug: {
                description: 'description',
                color: 'color',
              },
            },
          },
          additions: [
            {
              name: 'test',
              description: 'description-test',
              color: 'color',
              default: false,
            },
          ],
          updates: [
            {
              name: 'bug',
              description: 'description',
              color: 'color',
              default: false,
            },
          ],
          removals: [
            {
              name: 'enchantment',
              description: 'description',
              color: 'color',
              default: false,
            },
          ],
        },
        {
          name: 'graphql-shield',
          config: {
            strict: true,
            labels: {
              test: {
                description: 'description',
                color: 'color',
              },
              bug: {
                description: 'description',
                color: 'color',
              },
            },
          },
          additions: [
            {
              name: 'test',
              description: 'description',
              color: 'color',
              default: false,
            },
          ],
          updates: [
            {
              name: 'bug',
              description: 'description',
              color: 'color',
              default: false,
            },
          ],
          removals: [
            {
              name: 'enchantment',
              description: 'description',
              color: 'color',
              default: false,
            },
          ],
        },
        {
          name: 'github-labels',
          config: {
            labels: {
              test: {
                description: 'description',
                color: 'color',
              },
              bug: {
                description: 'description',
                color: 'color',
              },
            },
          },
          additions: [],
          updates: [],
          removals: [],
        },
        {
          name: 'github-labels',
          config: {
            labels: {
              test: {
                description: 'description',
                color: 'color',
              },
              bug: {
                description: 'description',
                color: 'color',
              },
            },
          },
          additions: [
            {
              name: 'test',
              description: 'description',
              color: 'color',
              default: false,
            },
            {
              name: 'test',
              description: 'description',
              color: 'color',
              default: false,
            },
          ],
          updates: [],
          removals: [],
        },
      ],
      errors: [
        {
          name: 'github-labels',
          config: {
            labels: {
              test: {
                description: 'description',
                color: 'color',
              },
              bug: {
                description: 'description',
                color: 'color',
              },
            },
          },
          message: 'Something horrific has occured.',
        },
      ],
    } as SyncReport)

    expect(report).toMatchSnapshot()
  })

  test('correctly builds empty organization sync report', async () => {
    const report = generateSyncReport({
      config: {},
      options: {
        githubToken: '',
        dryRun: true,
      },
      successes: [],
      errors: [],
    })

    expect(report).toMatchSnapshot()
  })
})
