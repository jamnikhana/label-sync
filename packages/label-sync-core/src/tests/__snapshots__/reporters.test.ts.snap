// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reporters correctly build report for organization sync 1`] = `
"
Github Labels Report
This is an autogenerated report for your project.

Hey ðŸ‘‹ we ran sync and this came out:

Problems


There were some issues with 1 repositories;

Couldn not sync github-labels:

Something horrific has occured.

### Configuration
\`\`\`json
{
  "labels": {
    "test": {
      "description": "description",
      "color": "123fff"
    },
    "bug": {
      "description": "description",
      "color": "123fff"
    }
  }
}\`\`\`
    
    

Changes


Successfully synced labels accross 4 repositories;

Synced github-labels:

Added 2 labels:
 - test
     description-test
 - simple
        

Updated 1 labels:
 - bug
     description
        

1 labels should be removed;
 - enchantment
     description

To remove them, set strict property to true.
    
    

Synced graphql-shield:

Added 1 labels:
 - test
     description
        

Updated 1 labels:
 - bug
     description
        

Removed 1 labels:
 - enchantment
     description
                  
    

Synced github-labels:
No new labels.
No labels updated.
No labels removed.
    

Synced github-labels:

Added 2 labels:
 - test
     description
 - test
     description
        
No labels updated.
No labels removed.
    
    
      "
`;

exports[`Reporters correctly builds empty organization sync report 1`] = `
"
Github Labels Report
This is an autogenerated report for your project.

Hey ðŸ‘‹ we ran sync and this came out:

Problems

Everything looks fine.

Changes

No successful sync reports.
      "
`;
